src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,289 @@
+
+    WebSearchHandler *webSearchHandler = m_searches[id];
+    if (!m_searches.contains(id))
+        throw APIError(APIErrorType::BadParams);
+
+    webSearchHandler->downloadTorrent(torrentUrl, siteUrl);
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
+    const QString name = params()["name"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    foreach (const QString &category, SearchPluginManager::instance()->getPluginCategories(name))
+        categories << SearchPluginManager::categoryFullName(category);
+
+    setResult(categories.join(","));
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    Logger *const logger = Logger::instance();
+
+    if (updateInfo.isEmpty()) {
+        logger->addMessage(tr("All plugins are already up to date."), Log::INFO);
+        return;
+    }
+
+    logger->addMessage(tr("Updating %1 plugins").arg(updateInfo.size()), Log::INFO);
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    for (const QString &pluginName : updateInfo.keys()) {
+        logger->addMessage(tr("Updating plugin %1").arg(pluginName), Log::INFO);
+        pluginManager->updatePlugin(pluginName);
+    }
+}
+
+void SearchController::checkForUpdatesFailed(const QString &reason)
+{
+    Logger::instance()->addMessage(tr("Failed to check for plugin updates: %1").arg(reason), Log::INFO);
+}
+
+QString SearchController::generateSearchId() const
```

Comment:
Actually, each search job is identified by SearchHandler instance so you can use its pointer as ID. Just serialize it.

Reply:
There are potential security implications to exposing the pointer's memory address to the WebUI. I prefer the current process of pseudo-random id generation.

Reply:
What about just using a counter and increment it every time?
I reckon this id doesn't really need to be strongly protected, or not?

Reply:
I don't think it needs to be strongly protected, since it's only useful once the user has authenticated. I considered the counter approach but that's an additional state we'd have to track.